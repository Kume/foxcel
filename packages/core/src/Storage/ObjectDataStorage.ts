import {DataStorage} from './DataStorage';

export class ObjectDataStorage implements DataStorage {
  get data(): {[key: string]: string} {
    return this._data;
  }

  private _data: {[key: string]: string} = {};

  public deleteHistory: (readonly string[])[] = [];
  public writeHistory: [readonly string[], string][] = [];
  public readHistory: (readonly string[])[] = [];

  public clearHistory(): void {
    this.deleteHistory = [];
    this.writeHistory = [];
    this.readHistory = [];
  }

  public async saveAsync(paths: Array<string>, content: string): Promise<void> {
    this.writeHistory.push([paths, content]);
    this._data[paths.join('/')] = content;
  }

  public async loadAsync(paths: Array<string>): Promise<string> {
    this.readHistory.push(paths);
    const path = paths.join('/');
    if (!(path in this._data)) {
      throw new Error(`File not found. ${path}`);
    }
    return this._data[path];
  }

  public async exists(paths: string[]): Promise<boolean> {
    return true;
  }

  public async deleteAsync(paths: Array<string>): Promise<void> {
    this.deleteHistory.push(paths);
    delete this._data[paths.join('/')];
  }
}
