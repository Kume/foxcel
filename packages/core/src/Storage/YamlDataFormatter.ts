import {DataFormatter} from './DataFormatter';
import * as Yaml from 'js-yaml';

export class YamlDataFormatter implements DataFormatter {
  format(data: unknown): string {
    return Yaml.dump(data);
  }

  parse(source: string): unknown {
    return Yaml.load(source);
  }
}
