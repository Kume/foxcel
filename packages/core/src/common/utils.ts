export function mapObjectToObject<V, R>(
  obj: {[key: string]: V},
  mapper: (value: V, key: string) => R,
): {readonly [key: string]: R} {
  const ret: {[key: string]: R} = {};
  for (const [key, value] of Object.entries(obj)) {
    ret[key] = mapper(value, key);
  }
  return ret;
}

export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R | undefined] | undefined,
  ignoreUndefined: true,
): {[key: string]: R};
export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R] | undefined,
  ignoreUndefined?: false,
): {[key: string]: R};
export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R] | undefined,
  ignoreUndefined?: boolean,
): {[key: string]: R} {
  const ret: {[key: string]: R} = {};
  for (let i = 0; i < values.length; i++) {
    const mapped = mapper(values[i], i);
    if (mapped) {
      const [key, value] = mapped;
      if (!ignoreUndefined || value !== undefined) {
        ret[key] = value;
      }
    }
  }
  return ret;
}

export function resolvePath(currentPath: readonly string[], targetPath: readonly string[]): string[] {
  const path = [...currentPath, ...targetPath];
  let i = 0;
  while (i < path.length) {
    if (path[i] === '..') {
      if (i === 0) {
        throw new Error('Cannot reference above the top directory');
      }
      path.splice(i - 1, 2);
      i--;
    } else {
      i++;
    }
  }
  return path;
}

export function validIndexOrUndefined(index: number): number | undefined {
  return index < 0 ? undefined : index;
}

export const emptyArray: readonly any[] = [];
Object.freeze(emptyArray);

export function emptyFunction(): void {
  // 何もしない
}

export function compact<T>(values: readonly (T | null | undefined | false)[]): T[] {
  return values.filter((i) => i !== undefined && i !== null && i !== false) as T[];
}

export function isReadonlyArray<T>(value: readonly T[] | unknown): value is readonly T[] {
  return Array.isArray(value);
}

export function rangeBySize(start: number, size: number): number[] {
  return [...Array(size)].map((_, i) => start + i);
}

export function rangeLt(start: number, sup: number): number[] {
  return rangeBySize(start, sup - start);
}

export function rangeInArrayEqual<Item>(
  a: readonly Item[],
  b: readonly Item[],
  itemEquals: (a: Item, b: Item) => boolean,
  start: number,
  size: number,
): boolean {
  for (let i = start; i < start + size; i++) {
    if (!itemEquals(a[i], b[i])) {
      return false;
    }
  }
  return true;
}

export function isNullOrEmptyString(value: string | undefined | null): value is '' | null | undefined {
  return value == null || value === '';
}

export function compareVersion(a: `${number}.${number}`, b: `${number}.${number}`): -1 | 0 | 1 {
  const [a1, a2] = a.split('.').map(Number);
  const [b1, b2] = b.split('.').map(Number);
  const sign1 = Math.sign(a1 - b1) as -1 | 0 | 1;
  if (sign1 !== 0) {
    return sign1;
  }
  return Math.sign(a2 - b2) as -1 | 0 | 1;
}

export function versionGte(a: `${number}.${number}`, b: `${number}.${number}`): boolean {
  return compareVersion(a, b) >= 0;
}
