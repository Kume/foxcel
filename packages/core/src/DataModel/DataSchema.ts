import {
  ConditionalDataSchemaItemConfig,
  ConditionalSchemaItem,
  ConditionConfig,
  DataPath,
  dataPathComponentIsKey,
  dataPathComponentIsPointer,
  DataSchemaConfig,
  DataSchemaConfigBase,
  ForwardDataPath,
  ForwardDataPathComponent,
  forwardDataPathComponentToString,
  RootSchemaConfig,
  SelectDynamicOptionConfig,
  SelectOptionConfig,
  SelectOptionConfigItem,
  SelectStaticOptionConfig,
} from '..';
import {MultiDataPath, parsePath} from './DataPath';
import {CommonReferenceSchema, FilePathConfigNamedItemMap, WritableFileBaseNamedItemNode} from '../common/commonTypes';
import DataStorage from '../Storage/DataStorage';
import {DataFormatter} from '../Storage/DataFormatter';
import {loadNestedConfigFile} from '../Storage/utils';
import {LoadedSchemaPath, resolveConfigOrRecursive} from '../common/schemaCommon';
import {dataPathToTemplateLine, parseTemplateLine, TemplateLine} from './TemplateEngine';

export enum DataSchemaType {
  Number,
  Boolean,
  String,
  Map,
  FixedMap,
  List,
  Reference,
  Conditional,
  Recursive,
  Key,
}

export type DataSchema =
  | NumberDataSchema
  | BooleanDataSchema
  | StringDataSchema
  | MapDataSchema
  | FixedMapDataSchema
  | ListDataSchema
  | ConditionalDataSchema
  | RecursiveDataSchema
  | KeyDataSchema;

export type DataSchemaExcludeRecursive<T = DataSchema> = T extends RecursiveDataSchema ? never : T;

type TypeForType<T, Type> = T extends {t: Type} ? T : never;

export type DataSchemaForType<T extends DataSchemaType> = TypeForType<DataSchema, T>;

export interface DataSchemaBase<T> {
  readonly t: DataSchemaType;
  readonly label?: string;
  readonly dataLabel?: TemplateLine;
  readonly dataDescription?: string;
  readonly required?: boolean;
  readonly defaultToUndefined?: boolean;
  readonly default?: T;
  readonly filePath?: readonly string[];
}

export interface NumberDataSchema extends DataSchemaBase<number> {
  readonly t: DataSchemaType.Number;
}

export interface BooleanDataSchema extends DataSchemaBase<boolean> {
  readonly t: DataSchemaType.Boolean;
}

export interface SelectStaticOptionSchema<T> {
  readonly label: string;
  readonly value: T;
}

export interface SelectDynamicOptionSchema {
  readonly label?: undefined;
  readonly path: MultiDataPath;
  readonly labelTemplate?: TemplateLine;
  readonly valuePath: ForwardDataPath | undefined;
}

export type SelectOptionSchema<T> = SelectStaticOptionSchema<T> | SelectDynamicOptionSchema;

export interface StringDataSchema extends DataSchemaBase<string> {
  readonly t: DataSchemaType.String;
  readonly in?: readonly SelectOptionSchema<string>[];
}

export interface MapDataSchema extends DataSchemaBase<unknown> {
  readonly t: DataSchemaType.Map;
  readonly contextKey?: string;
  readonly sourcePath?: DataPath;
  readonly item?: DataSchema;
}

export interface FixedMapDataSchema extends DataSchemaBase<{[key: string]: any}> {
  readonly t: DataSchemaType.FixedMap;
  readonly contextKey?: string;
  readonly items: {readonly [key: string]: DataSchema};
}

export interface ListDataSchema extends DataSchemaBase<any[]> {
  readonly t: DataSchemaType.List;
  readonly contextKey?: string;
  readonly item?: DataSchema;
}

export interface ReferenceDataSchema extends CommonReferenceSchema {
  readonly t: DataSchemaType.Reference;
}

export interface RecursiveDataSchema {
  readonly t: DataSchemaType.Recursive;
  readonly depth: number;
}

export interface ConditionalDataSchema {
  readonly t: DataSchemaType.Conditional;
  readonly label: string | undefined;
  readonly items: {
    readonly [key: string]: ConditionalSchemaItem<DataSchema, DataPath>;
  };
  readonly filePath?: readonly string[];
}

export interface KeyDataSchema {
  readonly t: DataSchemaType.Key;
  readonly label: string | undefined;
}

export class DataSchemaContext {
  public static createRootContext(rootSchema: DataSchemaExcludeRecursive | undefined): DataSchemaContext | undefined {
    return rootSchema && new DataSchemaContext(rootSchema, rootSchema, []);
  }

  private constructor(
    public readonly rootSchema: DataSchemaExcludeRecursive,
    public readonly currentSchema: DataSchemaExcludeRecursive,
    private readonly path: readonly DataSchemaExcludeRecursive[],
  ) {}

  public dig(key: {t: 'key'} | string | number | undefined | null): DataSchemaContext | undefined {
    if (key !== null && typeof key === 'object') {
      // TODO DataSchemaContextをkeyまで対応させるなら何かしら実装が必要
      return undefined;
    }

    switch (this.currentSchema.t) {
      case DataSchemaType.Map:
      case DataSchemaType.List:
        // MapとListはキー(インデックス)に関わらず子のスキーマは一つであるため、そのまま子を返す。
        return this.currentSchema.item && this.createFromNext(this.currentSchema.item);
      case DataSchemaType.FixedMap:
        if (typeof key === 'string') {
          return this.currentSchema.items[key] && this.createFromNext(this.currentSchema.items[key]);
        } else {
          return undefined;
        }
      case DataSchemaType.Conditional:
        if (typeof key === 'string') {
          return this.currentSchema.items[key] && this.createFromNext(this.currentSchema.items[key].item);
        } else {
          return undefined;
        }
      default:
        return undefined;
    }
  }

  public digByPath(pathComponent: ForwardDataPathComponent): DataSchemaContext | undefined {
    if (dataPathComponentIsKey(pathComponent)) {
      // TODO DataSchemaContextをkeyまで対応させるなら何かしら実装が必要
      return undefined;
    }
    switch (this.currentSchema.t) {
      case DataSchemaType.Map:
      case DataSchemaType.List:
        // MapとListはキー(インデックス)に関わらず子のスキーマは一つであるため、そのまま子を返す。
        return this.currentSchema.item && this.createFromNext(this.currentSchema.item);
      case DataSchemaType.FixedMap: {
        if (dataPathComponentIsPointer(pathComponent)) {
          // pointerタイプのpathはfixedMapの子を指すためのポインターとしては利用されない想定
          return undefined;
        }
        const key = forwardDataPathComponentToString(pathComponent);
        return this.currentSchema.items[key] && this.createFromNext(this.currentSchema.items[key]);
      }
      case DataSchemaType.Conditional:
        // TODO ここでconditionalを下降するためには実データがないと条件が定まらないので不可能?をもそもdataPathでdigするのが間違い?
        //  現状わからないので一旦undefinedを返す
        return undefined;
      default:
        return undefined;
    }
  }

  public getMapChild(): DataSchemaContext | undefined {
    const {currentSchema} = this;
    if (currentSchema.t !== DataSchemaType.Map) {
      return;
    }
    return currentSchema.item && this.createFromNext(currentSchema.item);
  }

  public getFixedMapChild(key: string): DataSchemaContext | undefined {
    const {currentSchema} = this;
    if (currentSchema.t !== DataSchemaType.FixedMap) {
      return;
    }
    return currentSchema.items[key] && this.createFromNext(currentSchema.items[key]);
  }

  public getListChild(): DataSchemaContext | undefined {
    const {currentSchema} = this;
    if (currentSchema.t !== DataSchemaType.List) {
      return;
    }
    return currentSchema.item && this.createFromNext(currentSchema.item);
  }

  private createFromNext(next: DataSchema): DataSchemaContext | undefined {
    const {rootSchema, path} = this;
    if (next.t === DataSchemaType.Recursive) {
      if (next.depth > path.length) {
        throw new Error('Invalid recursive depth');
      }
      return new DataSchemaContext(rootSchema, path[path.length - next.depth], path.slice(0, -next.depth));
    } else {
      return new DataSchemaContext(rootSchema, next, [...path, next]);
    }
  }

  public resolveRecursive(schema: DataSchema): DataSchemaExcludeRecursive;
  public resolveRecursive(schema: DataSchema | undefined): DataSchemaExcludeRecursive | undefined;
  public resolveRecursive(schema: DataSchema | undefined): DataSchemaExcludeRecursive | undefined {
    if (!schema) {
      return undefined;
    }
    if (schema.t === DataSchemaType.Recursive) {
      if (schema.depth > this.path.length) {
        throw new Error('Invalid recursive depth');
      }
      return this.path[this.path.length - schema.depth];
    } else {
      return schema;
    }
  }
}

function isDynamicOptionConfig<T>(
  config: SelectStaticOptionConfig<T> | SelectDynamicOptionConfig,
): config is SelectDynamicOptionConfig {
  return config.path !== undefined;
}

export function parseOptionConfig(config: SelectOptionConfig<string>): readonly SelectOptionSchema<string>[];
export function parseOptionConfig(
  config: SelectOptionConfig<string | number>,
): readonly SelectOptionSchema<string | number>[];
export function parseOptionConfig(
  config: SelectOptionConfig<string | number>,
): readonly SelectOptionSchema<string | number>[] {
  if (Array.isArray(config)) {
    return config.map(parseOptionConfigItem);
  } else {
    return [parseDynamicOptionConfigItem(config)];
  }
}

export function parseOptionConfigItem(config: SelectOptionConfigItem<string>): SelectOptionSchema<string>;
export function parseOptionConfigItem(
  config: SelectOptionConfigItem<string | number>,
): SelectOptionSchema<string | number>;
export function parseOptionConfigItem(
  config: SelectOptionConfigItem<string | number>,
): SelectOptionSchema<string | number> {
  if (typeof config === 'string' || typeof config === 'number') {
    return {label: config.toString(), value: config};
  }
  if (isDynamicOptionConfig(config)) {
    return parseDynamicOptionConfigItem(config);
  } else {
    return {...config};
  }
}

function parseDynamicOptionConfigItem(config: SelectDynamicOptionConfig): SelectDynamicOptionSchema {
  return {
    path: parsePath(config.path),
    labelTemplate: config.label
      ? parseTemplateLine(config.label)
      : config.labelPath
      ? dataPathToTemplateLine(parsePath(config.labelPath, 'single'))
      : undefined,
    valuePath: (config.valuePath && parsePath(config.valuePath, 'forward')) || undefined,
  };
}

export function parseCondition(source: ConditionConfig): ConditionConfig<DataPath> {
  if ('match' in source) {
    return {match: source.match, path: parsePath(source.path, 'single')};
  } else if ('or' in source) {
    return {or: source.or.map((i) => parseCondition(i))};
  } else if ('and' in source) {
    return {and: source.and.map((i) => parseCondition(i))};
  }
  throw new Error('Invalid condition');
}

export function parseConditions<DistItem>(
  sourceConditions: {[key: string]: ConditionalDataSchemaItemConfig},
  parseItem: (key: string, item: DataSchemaConfig | string) => DistItem,
): {readonly [key: string]: ConditionalSchemaItem<DistItem, DataPath>} {
  const conditions: {[key: string]: ConditionalSchemaItem<DistItem, DataPath>} = {};
  for (const key of Object.keys(sourceConditions)) {
    conditions[key] = {
      condition: parseCondition(sourceConditions[key].condition),
      item: parseItem(key, sourceConditions[key].item),
    };
  }
  return conditions;
}

function baseSchema<T extends DataSchemaType>(config: DataSchemaConfigBase, type: T) {
  return {
    t: type,
    label: config.label,
    dataLabel: config.dataLabel === undefined ? undefined : parseTemplateLine(config.dataLabel),
    dataDescription: config.dataDescription,
  };
}

function parseDataSchemaConfig(
  config: DataSchemaConfig,
  pathConfigMap: FilePathConfigNamedItemMap<DataSchemaConfig>,
  filePath: readonly string[] = [],
  loadedPath: LoadedSchemaPath = [],
): DataSchema {
  switch (config.type) {
    case 'number': {
      return {...baseSchema(config, DataSchemaType.Number)};
    }
    case 'boolean': {
      return {...baseSchema(config, DataSchemaType.Boolean)};
    }
    case 'fixed_map': {
      const items: {[key: string]: DataSchema} = {};
      for (const key of Object.keys(config.items)) {
        items[key] = parseChildDataSchemaConfig(config.items[key], pathConfigMap, filePath, loadedPath);
      }
      return {...baseSchema(config, DataSchemaType.FixedMap), items};
    }
    case 'list': {
      const item = parseChildDataSchemaConfig(config.item, pathConfigMap, filePath, loadedPath);
      return {...baseSchema(config, DataSchemaType.List), item};
    }
    case 'map': {
      const item = parseChildDataSchemaConfig(config.item, pathConfigMap, filePath, loadedPath);
      return {...baseSchema(config, DataSchemaType.Map), item};
    }
    case 'string': {
      const option = config.in && parseOptionConfig(config.in);
      return {...baseSchema(config, DataSchemaType.String), in: option};
    }
    case 'conditional': {
      const items = parseConditions(config.items, (key, item) =>
        parseChildDataSchemaConfig(item, pathConfigMap, filePath, loadedPath),
      );
      return {...baseSchema(config, DataSchemaType.Conditional), items};
    }
  }
}

function parseChildDataSchemaConfig(
  configOrReference: DataSchemaConfig | string,
  pathConfigMap: FilePathConfigNamedItemMap<DataSchemaConfig>,
  filePath: readonly string[],
  loadedPath: LoadedSchemaPath,
): DataSchema {
  const result = resolveConfigOrRecursive(configOrReference, pathConfigMap, filePath, loadedPath);
  if (result.type === 'recursive') {
    return {t: DataSchemaType.Recursive, depth: result.depth};
  } else {
    return parseDataSchemaConfig(result.config, pathConfigMap, result.filePath, result.loadedPath);
  }
}

export async function buildDataSchema(
  rootSchemaConfig: RootSchemaConfig,
  storage: DataStorage,
  formatter: DataFormatter,
): Promise<DataSchemaExcludeRecursive> {
  const namedDataSchema: WritableFileBaseNamedItemNode<DataSchemaConfig> = {filePath: []};
  const loadedDataSchema = new Map([['', namedDataSchema]]);
  await loadNestedConfigFile(
    rootSchemaConfig.namedDataSchema || {},
    namedDataSchema,
    loadedDataSchema,
    (config) => config, // TODO バリデーション
    storage,
    formatter,
  );
  return parseDataSchemaConfig(rootSchemaConfig.dataSchema, loadedDataSchema) as DataSchemaExcludeRecursive; // TODO
}

export function buildSimpleDataSchema(rootSchemaConfig: RootSchemaConfig): DataSchemaExcludeRecursive {
  const namedDataSchema: WritableFileBaseNamedItemNode<DataSchemaConfig> = {filePath: []};
  const loadedDataSchema = new Map([['', namedDataSchema]]);
  return parseDataSchemaConfig(rootSchemaConfig.dataSchema, loadedDataSchema) as DataSchemaExcludeRecursive; // TODO
}

export function getByKeyForFixedMapDataSchema(map: FixedMapDataSchema, key: string): DataSchema | undefined {
  return map.items[key];
}

export interface DataSchemaContextKeyItem {
  key: string;
  depth: number;
}

export function dataSchemaIsFixedMap(schema: DataSchema | undefined): schema is FixedMapDataSchema {
  return schema?.t === DataSchemaType.FixedMap;
}

export function dataSchemaIsMap(schema: DataSchema | undefined): schema is MapDataSchema {
  return schema?.t === DataSchemaType.Map;
}

export function dataSchemaIsList(schema: DataSchema | undefined): schema is ListDataSchema {
  return schema?.t === DataSchemaType.List;
}

export function dataSchemaIsString(schema: DataSchema | undefined): schema is StringDataSchema {
  return schema?.t === DataSchemaType.String;
}

export function dataSchemaIsNumber(schema: DataSchema | undefined): schema is NumberDataSchema {
  return schema?.t === DataSchemaType.Number;
}

export function dataSchemaIsBoolean(schema: DataSchema | undefined): schema is BooleanDataSchema {
  return schema?.t === DataSchemaType.Boolean;
}
