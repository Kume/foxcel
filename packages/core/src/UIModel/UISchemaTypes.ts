import {
  BooleanDataSchema,
  ConditionalDataSchema,
  FixedMapDataSchema,
  KeyDataSchema,
  ListDataSchema,
  MapDataSchema,
  NumberDataSchema,
  SelectOptionSchema,
  StringDataSchema,
} from '../DataModel/DataSchema';
import {DataPath} from '../DataModel/DataPath';
import type {UISchemaKey} from './UISchema';

interface UISchemaBase {
  readonly label?: string;
}

export type FlattenableUISchemaCommon =
  | {keyFlatten: true; flatKeys: ReadonlyMap<UISchemaKey, readonly UISchemaKey[]>}
  | {keyFlatten?: false};

export type UISchemaOrRecursive = UISchema | RecursiveUISchema;

export type UISchema =
  | MappingTableUISchema
  | CheckBoxUISchema
  | ContentListUISchema
  | FormUISchema
  | NumberUISchema
  | SelectUISchema
  | TableUISchema
  | TabUISchema
  | TextUISchema
  | ConditionalUISchema;

export interface MappingTableUISchema extends UISchemaBase {
  readonly type: 'mappingTable';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: MapDataSchema;
  readonly sourcePath: DataPath;
  readonly contents: ReadonlyArray<UISchemaOrRecursive>;
}

export interface TextUISchema extends UISchemaBase {
  readonly type: 'text';
  readonly key?: UISchemaKey;
  readonly keyFlatten?: undefined;
  readonly dataSchema: StringDataSchema | KeyDataSchema;
  readonly multiline: boolean | undefined;
}

export type TabUISchema = {
  readonly type: 'tab';
  readonly key?: UISchemaKey;
  readonly dataSchema: FixedMapDataSchema;
  readonly contents: ReadonlyArray<UISchemaOrRecursive>;
} & UISchemaBase &
  FlattenableUISchemaCommon;

export interface TableUISchema extends UISchemaBase {
  readonly type: 'table';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: ListDataSchema | MapDataSchema;
  readonly contents: ReadonlyArray<UISchemaOrRecursive>;
}

export type SelectUISchema = SingleSelectUISchema | MultiSelectUISchema;

interface SelectUISchemaBase extends UISchemaBase {
  readonly type: 'select';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly options: readonly SelectOptionSchema<string | number>[];
}

interface SingleSelectUISchema extends SelectUISchemaBase {
  readonly dataSchema: StringDataSchema | NumberDataSchema;
  readonly isMulti: false;
}

interface MultiSelectUISchema extends SelectUISchemaBase {
  readonly dataSchema: ListDataSchema;
  readonly isMulti: true;
}

export interface NumberUISchema extends UISchemaBase {
  readonly type: 'number';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: NumberDataSchema;
}

export type FormUISchema = {
  readonly type: 'form';
  readonly key?: UISchemaKey;
  readonly dataSchema: FixedMapDataSchema;
  readonly contents: ReadonlyArray<UISchemaOrRecursive>;
} & UISchemaBase &
  FlattenableUISchemaCommon;

export interface ContentListUISchema extends UISchemaBase {
  readonly type: 'contentList';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: ListDataSchema | MapDataSchema;
  readonly content: UISchemaOrRecursive;
}

export interface CheckBoxUISchema extends UISchemaBase {
  readonly type: 'checkbox';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: BooleanDataSchema;
}

export interface ConditionalUISchema extends UISchemaBase {
  readonly type: 'conditional';
  readonly key?: string;
  readonly keyFlatten?: undefined;
  readonly dataSchema: ConditionalDataSchema;
  readonly contents: {readonly [key: string]: UISchemaOrRecursive};
}

export interface RecursiveUISchema {
  readonly type: 'recursive';
  readonly depth: number;

  readonly key?: string;
  /**
   * 参照先のUISchemaに定義されているDataSchemaを使うべきなのでRecursiveだけdataSchemaを定義しないが、
   * optionalChaining中にアクセスする場合に型定義だけあった方が便利なのでnever型で明示的にプロパティの存在を定義。
   */
  readonly dataSchema?: never;
}
