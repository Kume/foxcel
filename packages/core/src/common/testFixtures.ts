import {RootSchemaConfig} from './ConfigTypes';
import {DataModel} from '../DataModel/DataModelTypes';
import {unknownToDataModel} from '../DataModel/DataModel';

export const configFixtures = {
  simpleRecursive: {
    schema: {
      dataSchema: {
        type: 'fixed_map',
        items: {
          root: 'recursiveItem',
        },
      },
      namedDataSchema: {
        recursiveItem: {
          type: 'list',
          label: 'Children',
          item: {
            type: 'fixed_map',
            items: {
              label: {
                type: 'string',
                label: 'Name',
              },
              children: 'recursiveItem',
            },
          },
        },
      },
      namedUiSchema: {
        recursiveUi: {
          key: 'root',
          type: 'contentList',
          content: {
            type: 'form',
            contents: [
              {type: 'text', key: 'label'},
              {type: 'ref', ref: 'recursiveUi', key: 'children'},
            ],
          },
        },
      },
      uiRoot: {
        type: 'tab',
        contents: [{type: 'ref', ref: 'recursiveUi', key: 'root'}],
      },
      fileMap: {children: []},
    },
    data: unknownToDataModel({
      root: [
        {
          children: [{children: []}],
        },
        {
          children: [],
        },
      ],
    }),
  },
} as const satisfies Readonly<Record<string, {readonly schema: RootSchemaConfig; readonly data?: DataModel}>>;
