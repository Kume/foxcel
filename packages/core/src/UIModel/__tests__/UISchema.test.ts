import {
  buildUISchema,
  createRootUiSchemaParsingContext,
  getUiSchemaUniqueKeyOrUndefined,
  parseUISchemaConfig,
} from '../UISchema';
import {UISchemaConfig} from '../..';
import {buildDataSchema, DataSchema, DataSchemaType} from '../../DataModel/DataSchema';
import {mapToObject} from '../../common/utils';
import {TabUISchema, UISchemaOrRecursive} from '../UISchemaTypes';
import {ObjectDataStorage} from '../../Storage/ObjectDataStorage';
import {YamlDataFormatter} from '../../Storage/YamlDataFormatter';
import {configFixtures} from '../../common/testFixtures';

const emptyRootContext = createRootUiSchemaParsingContext(undefined);

type TestData = {readonly uiSchema: UISchemaOrRecursive; readonly dataSchema: DataSchema};
type TestDataParam = {readonly key?: string};

const testDataCreator = {
  number: (): TestData => {
    const dataSchema = {
      t: DataSchemaType.Number,
      numberType: 'unsignedInteger',
      required: false,
      max: undefined,
      min: undefined,
    } as const;
    const uiSchema = {type: 'number', key: 'test_number', dataSchema} as const;
    return {uiSchema, dataSchema} as const;
  },
  text: (): TestData => {
    const dataSchema = {t: DataSchemaType.String, required: false} as const;
    const uiSchema = {type: 'text', key: 'test_text', dataSchema, multiline: undefined} as const;
    return {uiSchema, dataSchema} as const;
  },
  form: (param: TestDataParam, children: readonly TestData[]): TestData => {
    const dataSchema = {
      t: DataSchemaType.FixedMap,
      required: false,
      contextKey: undefined,
      pathAliases: undefined,
      items: mapToObject(children, (child) => {
        const key = getUiSchemaUniqueKeyOrUndefined(child.uiSchema);
        return key ? [key, child.dataSchema] : undefined;
      }),
    } as const;
    const uiSchema = {
      type: 'form',
      key: param.key!,
      contents: children.map((child) => child.uiSchema),
      dataSchema,
    } as const;
    return {uiSchema, dataSchema};
  },
  tab: (param: TestDataParam, children: readonly TestData[]): TestData => {
    const dataSchema = {
      t: DataSchemaType.FixedMap,
      required: false,
      contextKey: undefined,
      pathAliases: undefined,
      items: mapToObject(children, (child) => {
        const key = getUiSchemaUniqueKeyOrUndefined(child.uiSchema);
        return key ? [key, child.dataSchema] : undefined;
      }),
    } as const;
    const uiSchema: TabUISchema = {
      type: 'tab',
      key: param.key!,
      contents: children.map((child) => child.uiSchema),
      dataSchema,
    } as const;
    return {uiSchema, dataSchema};
  },
};

describe('Unit tests for parseUISchemaConfig', () => {
  describe('For text', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig({type: 'text'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('text');
      expect(result.dataSchema).toEqual({t: DataSchemaType.String});
    });
  });

  describe('For number', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig({type: 'number'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('number');
      expect(result.dataSchema).toEqual({t: DataSchemaType.Number});
    });
  });

  describe('For checkbox', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig({type: 'checkbox'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('checkbox');
      expect(result.dataSchema).toEqual({t: DataSchemaType.Boolean});
    });
  });

  describe('For form', () => {
    const configWithKeyFlatten: UISchemaConfig = {
      type: 'form',
      contents: [
        {type: 'text', key: 'test_text'},
        {type: 'form', key: 'test_form', contents: [{type: 'number', key: 'test_number'}]},
      ],
    };
    describe('Without DataSchema', () => {
      it('Simple', () => {
        const uiSchemaConfig: UISchemaConfig = {type: 'form', contents: [{type: 'text', key: 'test_text'}]};
        const result = parseUISchemaConfig(uiSchemaConfig, new Map(), emptyRootContext, undefined);
        expect(result).toEqual({
          type: 'form',
          contents: [testDataCreator.text().uiSchema],
          dataSchema: {
            t: DataSchemaType.FixedMap,
            items: {test_text: {t: DataSchemaType.String}},
          },
        });
      });

      it('Nested without keyFlatten', () => {
        const result = parseUISchemaConfig(configWithKeyFlatten, new Map(), emptyRootContext, undefined);
        const expected = testDataCreator.form({}, [
          testDataCreator.text(),
          testDataCreator.form({key: 'test_form'}, [testDataCreator.number()]),
        ]);
        expect(result).toEqual(expected.uiSchema);
        expect(result).toMatchSnapshot();
      });
    });

    describe('With DataSchema', () => {
      it('Nested without keyFlatten', () => {
        const expected = testDataCreator.form({}, [
          testDataCreator.text(),
          testDataCreator.form({key: 'test_form'}, [testDataCreator.number()]),
        ]);
        const result = parseUISchemaConfig(configWithKeyFlatten, new Map(), emptyRootContext, expected.dataSchema);
        expect(result).toEqual(expected.uiSchema);
        expect(result).toMatchSnapshot();
      });
    });
  });
});

describe('Unit tests for buildUISchema', () => {
  it('Recursive', async () => {
    const storage = new ObjectDataStorage();
    const fixture = configFixtures.simpleRecursive;
    const dataSchema = await buildDataSchema(fixture.schema, storage, new YamlDataFormatter());
    const uiSchema = await buildUISchema(fixture.schema, dataSchema, storage, new YamlDataFormatter());
    // TODO expect書く
    console.log(uiSchema);
  });
});
