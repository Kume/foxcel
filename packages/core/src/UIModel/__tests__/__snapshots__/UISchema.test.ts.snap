// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for parseUISchemaConfig For form With DataSchema Nested without keyFlatten 1`] = `
{
  "contents": [
    {
      "dataSchema": {
        "t": 2,
      },
      "key": "test_text",
      "multiline": undefined,
      "type": "text",
    },
    {
      "contents": [
        {
          "dataSchema": {
            "t": 0,
          },
          "key": "test_number",
          "type": "number",
        },
      ],
      "dataSchema": {
        "items": {
          "test_number": {
            "t": 0,
          },
        },
        "t": 4,
      },
      "key": "test_form",
      "label": undefined,
      "type": "form",
    },
  ],
  "dataSchema": {
    "items": {
      "test_form": {
        "items": {
          "test_number": {
            "t": 0,
          },
        },
        "t": 4,
      },
      "test_text": {
        "t": 2,
      },
    },
    "t": 4,
  },
  "key": undefined,
  "label": undefined,
  "type": "form",
}
`;

exports[`Unit tests for parseUISchemaConfig For form Without DataSchema Nested without keyFlatten 1`] = `
{
  "contents": [
    {
      "dataSchema": {
        "t": 2,
      },
      "key": "test_text",
      "multiline": undefined,
      "type": "text",
    },
    {
      "contents": [
        {
          "dataSchema": {
            "t": 0,
          },
          "key": "test_number",
          "type": "number",
        },
      ],
      "dataSchema": {
        "items": {
          "test_number": {
            "t": 0,
          },
        },
        "t": 4,
      },
      "key": "test_form",
      "label": undefined,
      "type": "form",
    },
  ],
  "dataSchema": {
    "items": {
      "test_form": {
        "items": {
          "test_number": {
            "t": 0,
          },
        },
        "t": 4,
      },
      "test_text": {
        "t": 2,
      },
    },
    "t": 4,
  },
  "key": undefined,
  "label": undefined,
  "type": "form",
}
`;
